### PR TITLE
P1899R3 stride_view

### LINKED ISSUE #5597: [2022-07 LWG Motion 10] P1899R3 stride_view
[P1899R3](https://wg21.link/p1899r3)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5597
Fixes cplusplus/papers#647

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -12881,3 +12891,615 @@
...
+constexpr range_difference_t<V> stride() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\exposid{stride_}.
+\end{itemdescr}
+
+\indexlibrarymember{stride}{stride_view}%
```
Comment (JohelEGP) [inline source/ranges.tex:13004]:
```suggestion
\indexlibrarymember{size}{stride_view}%
```

File: source/ranges.tex [hunk 2]
```diff
@@ -12881,3 +12891,615 @@
...
+\begin{itemdecl}
+constexpr auto size() requires sized_range<V>;
+constexpr auto size() const requires sized_range<const V>;
```
Comment (JohelEGP) [inline source/ranges.tex:13007]:
```suggestion
constexpr auto size() requires @\libconcept{sized_range}@<V>;
constexpr auto size() const requires @\libconcept{sized_range}@<const V>;
```

File: source/ranges.tex [hunk 3]
```diff
@@ -12881,3 +12891,615 @@
...
+
+    constexpr @\exposid{iterator}@(@\exposid{Parent}@* parent, iterator_t<@\exposid{Base}@> current,        // \expos
+                       range_difference_t<@\exposid{Base}@> missing = 0);
+  public:
+    using difference_type = range_difference_t<@\exposid{Base}@>;
+    using value_type = range_value_t<@\exposid{Base}@>;
+    using iterator_concept = @\seebelow@;
+    using iterator_category = @\seebelow@;    // not always present
+
+    iterator() requires default_initializable<iterator_t<@\exposid{Base}@>> = default;
```
Comment (JohelEGP) [inline source/ranges.tex:13043]:
```suggestion
    iterator() requires @\libconcept{default_initializable}@<iterator_t<@\exposid{Base}@>> = default;
```

File: source/ranges.tex [hunk 4]
```diff
@@ -12881,3 +12891,615 @@
...
+
+    constexpr @\exposid{iterator}@(@\exposid{iterator}@<!Const> other)
+      requires Const && convertible_to<iterator_t<V>, iterator_t<@\exposid{Base}@>>
+                     && convertible_to<sentinel_t<V>, sentinel_t<@\exposid{Base}@>>;
```
Comment (JohelEGP) [inline source/ranges.tex:13047]:
```suggestion
      requires Const && @\libconcept{convertible_to}@<iterator_t<V>, iterator_t<@\exposid{Base}@>>
                     && @\libconcept{convertible_to}@<sentinel_t<V>, sentinel_t<@\exposid{Base}@>>;
```

File: source/ranges.tex [hunk 5]
```diff
@@ -12881,3 +12891,615 @@
...
+    constexpr @\exposid{iterator}@& operator-=(difference_type n) requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
+
+    constexpr decltype(auto) operator[](difference_type n) const
+      requires @\libconcept{random_access_range}@<@\exposid{Base}@>
+    { return *(*this + n); }
+
+    friend constexpr bool operator==(const @\exposid{iterator}@& x, default_sentinel_t);
+
+    friend constexpr bool operator==(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+      requires equality_comparable<iterator_t<@\exposid{Base}@>>;
```
Comment (JohelEGP) [inline source/ranges.tex:13071]:
```suggestion
      requires @\libconcept{equality_comparable}@<iterator_t<@\exposid{Base}@>>;
```

File: source/ranges.tex [hunk 6]
```diff
@@ -12881,3 +12891,615 @@
...
+If \tcode{C} models
+\tcode{\libconcept{derived_from}<random_access_iterator_tag>},
+then \tcode{iterator_category} denotes \tcode{ran\-dom_access_iterator_tag}.
+\item
+Otherwise, \tcode{iterator_category} denotes \tcode{C}.
+\end{itemize}
+
+\indexlibraryctor{stride_view::iterator}%
+\begin{itemdecl}
+constexpr @\exposid{iterator}@(Parent* parent, iterator_t<@\exposid{Base}@> current,
```
Comment (JohelEGP) [inline source/ranges.tex:13147]:
```suggestion
constexpr @\exposid{iterator}@(@\exposid{Parent}@* parent, iterator_t<@\exposid{Base}@> current,
```

File: source/ranges.tex [hunk 7]
```diff
@@ -12881,3 +12891,615 @@
...
+
+\indexlibrarymember{operator++}{stride_view::iterator}%
+\begin{itemdecl}
+constexpr @\exposid{iterator}@& operator++();
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\exposid{current_} != \exposid{end_} is \tcode{true}.
```
Comment (JohelEGP) [inline source/ranges.tex:13206]:
```suggestion
\tcode{\exposid{current_} != \exposid{end_}} is \tcode{true}.
```

File: source/ranges.tex [hunk 8]
```diff
@@ -12881,3 +12891,615 @@
...
+constexpr @\exposid{iterator}@& operator--() requires @\libconcept{bidirectional_range}@<@\exposid{Base}@>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+ranges::advance(current_, missing_ - stride_);
+missing_ = 0;
```
Comment (JohelEGP) [inline source/ranges.tex:13255]:
```suggestion
ranges::advance(@\exposid{current_}@, @\exposid{missing_}@ - @\exposid{stride_}@);
@\exposid{missing_}@ = 0;
```

File: source/ranges.tex [hunk 9]
```diff
@@ -12881,3 +12891,615 @@
...
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+If \tcode{n} is positive,
+\tcode{ranges::distance(\exposid{current_}, \exposid{end_}) > \exposid{stride_} * (n - 1)}
+is \tcode{true}.
+\begin{note}
+If \tcode{n} is negative, the \effects paragraph implies a precondition.
```
Comment (JohelEGP) [inline source/ranges.tex:13288]:
```suggestion
If \tcode{n} is negative, the \Fundescx{Effects} paragraph implies a precondition.
```

File: source/ranges.tex [hunk 10]
```diff
@@ -12881,3 +12891,615 @@
...
+\indexlibrarymember{operator<=>}{stride_view::iterator}%
+\begin{itemdecl}
+friend constexpr auto operator<=>(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+  requires @\libconcept{random_access_range}@<@\exposid{Base}@> && @\libconcept{three_way_comparable}@<iterator_t<@\exposid{Base}@>>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{x.current_ <=> y.current_}.
```
Comment (JohelEGP) [inline source/ranges.tex:13397]:
```suggestion
\tcode{x.\exposid{current_} <=> y.\exposid{current_}}.
```

File: source/ranges.tex [hunk 11]
```diff
@@ -12881,3 +12891,615 @@
...
+\indexlibrarymember{operator-}{stride_view::iterator}%
+\begin{itemdecl}
+friend constexpr difference_type operator-(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+  requires @\libconcept{sized_sentinel_for}@<iterator_t<@\exposid{Base}@>, iterator_t<@\exposid{Base}@>>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+Let \tcode{N} be \tcode{x.\exposid{current_} - y.\exposid{current_}}.
```
Comment (JohelEGP) [inline source/ranges.tex:13445]: `-N` is used below. Though I'm not sure if this makes a difference.
```suggestion
Let \tcode{N} be \tcode{(x.\exposid{current_} - y.\exposid{current_})}.
```

File: source/ranges.tex [hunk 12]
```diff
@@ -12881,3 +12891,615 @@
...
+friend constexpr void iter_swap(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+  noexcept(noexcept(ranges::iter_swap(x.@\exposid{current_}@, y.@\exposid{current_}@)))
+  requires @\libconcept{indirectly_swappable}@<iterator_t<@\exposid{Base}@>>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to:
+\tcode{ranges::iter_swap(x.\exposid{current_}, y.\exposid{current_});}
```
Comment (JohelEGP) [inline source/ranges.tex:13504]:
```suggestion
\tcode{ranges::iter_swap(x.\exposid{current_}, y.\exposid{current_})}
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex (hunk 12 above)
Comment (jensmaurer) [reply source/ranges.tex:13504]: That's a "statement"-level "equivalent to", and the semicolon is part of the statement.

Comment (JohelEGP) [reply source/ranges.tex:13506]: They should be equivalent. And I forgot to add the period to my suggestion.

File: source/ranges.tex [hunk 13]
```diff
@@ -12882,3 +12892,616 @@
...
+
+    friend constexpr auto operator<=>(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y)
+        requires @\libconcept{random_access_range}@<@\exposid{Base}@> && @\libconcept{three_way_comparable}@<iterator_t<@\exposid{Base}@>>;
+
+    friend constexpr @\exposid{iterator}@& operator+(const @\exposid{iterator}@& x, difference_type n)
+      requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
+    friend constexpr @\exposid{iterator}@& operator+(difference_type n, const @\exposid{iterator}@& x)
+      requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
+    friend constexpr @\exposid{iterator}@& operator-(const @\exposid{iterator}@& x, difference_type n)
+      requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
```
Comment (CaseyCarter) [inline source/ranges.tex:13095]: Error in the ingress paper: All three of these should have a return type of `\exposid{iterator}`, _not_ `\exposid{iterator}&`. The return types are correct in the detailed specifications below.

Comment (JohelEGP) [reply source/ranges.tex:13034]:
```suggestion
    friend constexpr @\exposid{iterator}@ operator+(const @\exposid{iterator}@& x, difference_type n)
      requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
    friend constexpr @\exposid{iterator}@ operator+(difference_type n, const @\exposid{iterator}@& x)
      requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
    friend constexpr @\exposid{iterator}@ operator-(const @\exposid{iterator}@& x, difference_type n)
      requires @\libconcept{random_access_range}@<@\exposid{Base}@>;
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex (hunk 12 above)
Comment (tkoeppe) [reply source/ranges.tex:13451]: I think we allow expressions of void type to be used in sentence style, "Equivalent to `foo()`.", but I don't think we _have_ to do that, especially if there are a lot of similar nearby cases of non-void expressions where we want the colon form. I think it's fine to leave as is unless someone can make a strong case for why we should use sentence style.

**Pull request merged**